### PR TITLE
Adjustments to cluster-bot to support step based periodics

### PR DESCRIPTION
**Pull request by jupierce**
1. Post-steps do not run if `stopJob` deletes the entire project.
Changed to terminate ci-operator pod, which triggers post-steps to
run and for templates to gracefully teardown.
2. Step based jobs follow a pod/container naming convention different than
the traditional template jobs. Instead trying to reflect this, step based
jobs will monitor a secret in the test namespace to extract
cluster information. Access to this secret must be granted by the
openshift-cluster-bot-rbac registry step.

### CODE REVIEW

File: manager.go [hunk 1]
```diff
@@ -360,7 +360,7 @@ func (m *jobManager) sync() error {
 			j.State = prowapiv1.PendingState
 			j.Failure = ""
 
-			if j.Mode == "launch" {
+			if j.Mode == "launch" && (previous != nil && !previous.Complete) {
```
Comment (smarterclayton) [inline manager.go:363]: Good catch (can you put this in its own commit)?

File: prow.go [hunk 2]
```diff
@@ -50,7 +50,7 @@ var supportedUpgradeTests = []string{"e2e-upgrade", "e2e-upgrade-all", "e2e-upgr
 
 // supportedPlatforms requires a job within the release periodics that can launch a
 // cluster that has the label job-env: platform-name.
-var supportedPlatforms = []string{"aws", "gcp", "azure", "vsphere", "metal"}
+var supportedPlatforms = []string{"aws", "gcp", "azure", "vsphere", "metal", "osd-aws"}
```
Comment (smarterclayton) [inline prow.go:53]: hrm, i think osd should just be a platform, and you should be able to say `ows,osd` or just `osd` if you want it.

File: prow.go [hunk 3]
```diff
@@ -222,6 +236,9 @@ func (m *jobManager) newJob(job *Job) error {
 		if len(input.Version) == 0 {
 			continue
 		}
+		// OSD needs to resolve the pullspec itself from cluster version, so include it
+		// in the env vars.
+		prow.SetJobEnvVar(&pj.Spec, "CLUSTER_VERSION", input.Version)
```
Comment (smarterclayton) [inline prow.go:241]: ugh, this is a net new var i wasn't aware of.  I would probably prefer it be named differently - is only OSD using it right now?

Comment (jupierce) [reply prow.go:241]: Yes - It is only for OSD atm. Their API works based on published Cincinnati versions and can't accept an arbitrary sha. I kept the name generic because I thought future providers might have a similar limitation (e.g. cluster pools), but I'm certainly open to making it OSD specific. `OSD_CLUSTER_VERSION` ?

File: prow.go (hunk 2 above)
Comment (jupierce) [reply prow.go:53]: Perhaps:
```
var supportedPlatforms = []string{"aws", ..., "osd"}
var supportedParameters = []string{"ovn", "proxy", "compact", "fips", ..., "osd-aws"}
```
This would allow a logic branch in the `osd-create-create` step to handle the variants versus a new (nearly identical) periodic if `aws,osd` is a separate item in `supportedPlatforms`.

File: manager.go (hunk 1 above)
Comment (jupierce) [reply manager.go:363]: Extracted into first commit.

File: prow.go (hunk 2 above)
Comment (jupierce) [reply prow.go]: Current commit has `osd` as a platform and `osd-aws` as a variant (anticipating variants like `osd-gcp` in the future).

File: prow.go (hunk 3 above)
Comment (smarterclayton) [reply prow.go:241]: OSD_INSTALL_CLUSTER_VERSION for now would be ideal, with a big TODO: "this needs to be reconciled with a longer term strategy".

dedicated needs to work with images eventually.  Can you verify with Jeremy that's just a point in time thing?

File: prow.go [hunk 4]
```diff
@@ -945,14 +1051,23 @@ func findTargetName(spec *corev1.PodSpec) (string, error) {
...
 					value := (&resolvedEnvironment{env: container.Env}).Resolve(value)
 					if len(value) == 0 {
-						return "", fmt.Errorf("bug in resolving %s", value)
+						return "", false, fmt.Errorf("bug in resolving %s", value)
 					}
-					return value, nil
+					targetName = value
 				}
 			}
+			if strings.HasPrefix(arg, "--template") {
```
Comment (smarterclayton) [inline prow.go:1059]: This feels like a boolean that should be returned from this method or passed in.  Probably passed in.

File: prow.go (hunk 3 above)
Comment (jupierce) [reply prow.go]: The recommendation has been implemented in https://github.com/openshift/ci-chat-bot/pull/131 . References to OSD have been removed from this PR.

File: prow.go (hunk 4 above)
Comment (jupierce) [reply prow.go:1052]: The result of this check is returned as a boolean by the method. To change to "passed in", are you suggesting a method signature change to something like:
```
func findTarget(spec *corev1.PodSpec, isTemplate *bool) (string, error)
     ...
```
and setting `*isTemplate` in the body?

Comment (smarterclayton) [reply prow.go:1052]: no, you should be telling findTarget what type of job this is.  You should know up front whether this is template or step, not guess.  Then you should either have two methods (if necessary) or one method.  But the end goal is not accidentally mistaking one job for another because someone changed something.  THe step vs template determination needs to be done one time at the top of the job, not later.

File: prow.go [hunk 5]
```diff
@@ -130,12 +130,26 @@ func (r *URLConfigResolver) Resolve(org, repo, branch, variant string) ([]byte,
 	}
 }
 
-// stopJob triggers namespace deletion, which will cause graceful shutdown of the
-// cluster. If this method returns nil, it is safe to consider the cluster released.
+// stopJob triggers graceful cluster teardown. If this method returns nil,
+// it is safe to consider the cluster released.
 func (m *jobManager) stopJob(name string) error {
```
Comment (stevekuznetsov) [inline prow.go:135]: In both cases you should just delete the Pod running `ci-operator` in the appropriate `ci` namespace

File: prow.go [hunk 6]
```diff
@@ -557,41 +569,89 @@ func (m *jobManager) waitForJob(job *Job) error {
...
-			return false, errJobCompleted
-		}
-		pod, err := m.coreClient.CoreV1().Pods(namespace).Get(context.TODO(), targetPodName, metav1.GetOptions{})
-		if err != nil {
-			// pod could not be created or we may not have permission yet
-			if !errors.IsNotFound(err) && !errors.IsForbidden(err) {
-				lastErr = err
+		if stepBasedMode {
+			if m.jobIsComplete(job) {
+				// The goal here is to interrupt the installation pod or any other step involved in
```
Comment (stevekuznetsov) [inline prow.go:583]: Again you should be able to just delete the Pod running `ci-operator`, it propagates down

File: prow.go (hunk 5 above)
Comment (jupierce) [reply prow.go:135]: Thanks @stevekuznetsov . Verified this worked for both templates and steps and dramatically simplifies things. Added a new commit with the change.

Comment (jupierce) [reply prow.go:135]: After trying this at various points of the install, discovered https://github.com/openshift/release/pull/16552 where data written into SHARED_DIR is not available to the post pod. Will need to track this down before adopting this approach.

File: prow.go [hunk 7]
```diff
@@ -569,41 +584,67 @@ func (m *jobManager) waitForJob(job *Job) error {
 		return fmt.Errorf("unable to check launch status: %v", err)
 	}
 
-	klog.Infof("Job %q waiting for setup container in pod %s/%s to complete", job.Name, namespace, targetPodName)
+	if stepBasedMode {
+		klog.Infof("Job %q waiting for step secret %s/%s to be populated", job.Name, namespace, targetName)
```
Comment (smarterclayton) [inline prow.go:588]: I don't think you have to split this message - it doesn't add anything

File: prow.go [hunk 8]
```diff
@@ -955,12 +1033,17 @@ func findTargetName(spec *corev1.PodSpec) (string, error) {
...
+					targetName = value
 				}
 			}
 		}
 	}
-	return "", fmt.Errorf("could not find argument --target=X in prow job pod spec to identify target pod name")
+
+	if len(targetName) > 0 {
+		return targetName, nil
+	} else {
```
Comment (smarterclayton) [inline prow.go:1044]: don't else when the if returns

Comment (smarterclayton) [reply prow.go:1044]: but return errors early

Comment (jupierce) [reply prow.go]: All changes to the method reverted. They were no longer necessary with the change to determining step mode.

File: prow.go (hunk 5 above)
Comment (jupierce) [reply prow.go:135]: Fixed with workaround for https://bugzilla.redhat.com/show_bug.cgi?id=1937523: 
- https://github.com/openshift/ci-tools/pull/1805 
- https://github.com/openshift/ci-tools/pull/1823

**Comment by jupierce** [comment]
A new commit added which simplifies the stopJob implementation now that https://bugzilla.redhat.com/show_bug.cgi?id=1937523 is being worked around in ci-operator in production.

File: manager.go [hunk 9]
```diff
@@ -1185,24 +1194,26 @@ func (m *jobManager) TerminateJobForUser(user string) (string, error) {
...
 		return "", err
 	}
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
 	klog.Infof("user %q requests job %q to be terminated", user, name)
-	if err := m.stopJob(name); err != nil {
-		klog.Errorf("unable to terminate running cluster %s: %v", name, err)
+	if job, ok := m.jobs[name]; ok {
```
Comment (smarterclayton) [inline manager.go:1202]: putting stopjob under the lock violates the "no long running operations under the lock".  You probably need to refactor this to:

1. get the current job status under lock, then release
2. stop the job
3. if stopping is successful, take the log and update the job status

for 1 that means moving the "get job, check status" to a method with lock and defer, then here check the value.  you also want to return an error if stopErr errors out so that the user knows stop request failed.  that means then that we'll never "drop the request" until the stop job passes, but that also may mean we need additional logic here to clean up the state (i.e. you'll need to handle "job doesn't exist" and race conditions)

Comment (jupierce) [reply manager.go:1209]: ack on the time spent in the lock. I put `stopJob` inside the lock initially because it required the actual `job` metadata to determine if it was stopping a template or step job. Now that those paths share the same logic, `stopJob` does not require the `job` at all -- so I've moved it back to its original location outside the lock..
I did, however, close one race condition (`stopJob` run before prowjob is `Pending`) and return a message back to the user as requested.

File: prow.go [hunk 10]
```diff
@@ -130,14 +130,39 @@ func (r *URLConfigResolver) Resolve(org, repo, branch, variant string) ([]byte,
...
-// stopJob triggers namespace deletion, which will cause graceful shutdown of the
-// cluster. If this method returns nil, it is safe to consider the cluster released.
+// stopJob triggers graceful cluster teardown. If this method returns nil,
+// it is safe to consider the cluster released.
 func (m *jobManager) stopJob(name string) error {
-	namespace := fmt.Sprintf("ci-ln-%s", namespaceSafeHash(name))
-	if err := m.projectClient.ProjectV1().Projects().Delete(context.TODO(), namespace, metav1.DeleteOptions{}); err != nil && !errors.IsNotFound(err) {
+	uns, err := m.prowClient.Namespace(m.prowNamespace).Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		if !errors.IsNotFound(err) {
```
Comment (smarterclayton) [inline prow.go:138]: ?  This looks wierd.

Comment (smarterclayton) [reply prow.go:138]: Hrm...

Comment (jupierce) [reply prow.go:138]: Thanks for catching that. Removed `!` .

File: prow.go [hunk 11]
```diff
@@ -130,14 +130,39 @@ func (r *URLConfigResolver) Resolve(org, repo, branch, variant string) ([]byte,
...
+			// There may have been an issue creating the prowjob; treat as success
+			return nil
```
Comment (smarterclayton) [inline prow.go:140]: It still looks wierd - if we can't find the pod, we return "err can't find pod".  If get any other error, we return nil (but what error are you expecting here that would imply "we can go ahead and release it)?  User needs to retry on any error except NotFound.

Comment (jupierce) [reply prow.go:140]: Sorry, I have to validate on my test branch which enables osd & I accidentally did a relevant push there. For real this time, removing `!` .

**Comment by jupierce** [comment]
Testing looks good.
/unhold
